compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -287,6 +287,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expr_ty: Ty<'tcx>,
     ) {
         if let ty::Adt(expected_adt, substs) = expected.kind() {
+            if let hir::ExprKind::Field(base, ident) = expr.kind
+                && let Some(typeck_results) = self.in_progress_typeck_results
+            {
+                let base_ty = typeck_results.borrow().expr_ty(base);
+                if base_ty == expected {
```

Comment:
Can we use something like `can_eq` instead?

```suggestion
                if self.can_eq(self.param_env, base_ty, expected).is_ok() {
```

Reply:
alternatively we could use `same_type_modulo_regions`, but since we have access to `can_eq`, it's a bit more accurate.